### ISSUE TITLE
r10k failing git clone when branch directory exists

### ISSUE THREAD

**Issue by terrimonster** [puppetlabs/r10k#170]
```
/opt/puppet/bin/r10k deploy environment mybranch -p --verbose
[R10K::Source::Git - INFO] Determining current branches for "git@gitserver.com:puppet/puppet-control.git"
[R10K::Task::Environment::Deploy - NOTICE] Deploying environment mybranch
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Environment::Deploy:0x000000014860b0> failed while running: Command git clone --reference /var/cache/r10k/git@gitserver.com-puppet-puppet-control.git git@gitserver.com:puppet/puppet-control.git /etc/puppetlabs/puppet/environments/mybranch exited with 128: fatal: destination path '/etc/puppetlabs/puppet/environments/mybranch already exists and is not an empty directory.
```

Clearing out /var/cache/r10k did not fix this issue. rm -rf the branch in the environments dir did fix it.

puppet enterprise v. 3.3.0
centos 6.4
r10k v. 1.3.1 (installed via gem, managed via zack/r10k module)
using gitlab free version

**Comment by adrienthebo** [comment]
Could you provide a sanitized copy of the `r10k.yaml` that was used here? What are the contents of `/etc/puppetlabs/puppet/environments`?

**Comment by vchepkov** [comment]
I suspect I have a related problem.
When I use r10k version 1.1.0, all my environments are working copies of existing branches and git status reports, for example:
On branch development
Your branch is ahead of 'origin/development' by 195 commits.
  (use "git push" to publish your local commits)

Why it reports "ahead" I don't really understand, but it's current 

With r10k v1.3.2 I get detached heads in all my branches/environments and they are not up to date

Nothing special about my config

```
cat /etc/r10k.yaml 
:cachedir: /var/cache/r10k
:sources:
  puppet:
    basedir: /etc/puppetlabs/puppet/environments
    remote: git@stash.bna.com:/bp/puppet-bna.git

:purgedirs:
  - /etc/puppetlabs/puppet/environments

```

**Comment by adrienthebo** [comment]
@vchepkov r10k 1.2.0 switched from tracking branches in a semi-sloppy fashion to checking out exact refs, which is more correct and less strange but also less user friendly. GH-97 is tracking that issue.

@terrimonster is this still an issue? In order to diagnose this issue I need a bit more information about the configuration that produced this issue.

**Comment by adrienthebo** [comment]
This issue has been open for several months and I still need more information; if this is still a problem let me know, otherwise I'll close it out in a week or so.

**Comment by jameskyle** [comment]
Heya,

I'm seeing this behavior with same error in the following case
- Previous environments exist
- One of the previous environments has a name conflict with a branch in a managed environment

For example, say I have a branch called 'production' and a directory exists in the basedir by the same name.

**Comment by cvquesty** [comment]
I am experiencing the same error.  

If at provisioning time this occurs, you can rm -rf /etc/puppetlabs/puppet/environments/production and then r10k deploy environment production and it deploys just fine.  Previous versions of Puppet worked for me, but this issue only arose after moving to PE3.7

I do note that I found this in a Vagrantfile instance after directory environments have been configured.  Previously, no directory environments were there, and PE did not create this directory.  Now, the production environment pre-exists, and causes the issue with Vagrant automated host provisioning.

Anxiously awaiting this fix!!!!!

**Comment by ggeldenhuis** [comment]
I am experiencing the same issue.
The puppet install is a fresh copy of PE 3.7.1 on CentOS 6.6
Git server is gitlab.
I manually installed the zack/r10k module into /etc/puppetlabs/puppet/modules/
Then ran puppet apply against: 
class { 'r10k':
  remote => 'git@company.com:enterprise-provisioning/r10k-control.git',
}

Which gives the error:
fatal: destination path '/etc/puppetlabs/puppet/environments/production' already exists and is not an empty directory.
Exit code: 128

I can recreate the problem reliably.

**Comment by adrienthebo** [comment]
@gjngeldenhuis this occurs when an environment already exists and was not created by r10k, correct?

**Comment by ggeldenhuis** [comment]
@adrienthebo that is correct, pre-existing content will cause a run failure. PE 3.7.1 pre-populates the production directory with site.pp. I have place that site.pp in my r10k repo deleted the contents of the directory and were then able to run r10k without issue.

**Comment by adrienthebo** [comment]
Okay, then that's consistent with the previous reports. This is a valid issue, and hopefully will be fixed as part of 1.5.0.

**Comment by cvquesty** [comment]
Thanks for the attention, Adrien.  

What I've done (in Vagrant) is move the existing directory to a -BAK directory in the event someone has existing modules there. 

I know you know this, but I'd caution backing up what's there already before moving or deleting it, or massaging Git in such a way (this _is_ a Git issue, after all) that it is able to checkout into the modules folder without issue.  Last thing a user needs is to have their modules all whacked due to new behavior in r10k and their lack of planning.

**Comment by adrienthebo** [comment]
This issue should have been resolved in r10k 1.5.0, can anyone confirm that this is fixed? If I don't hear back I'll assume the issue resolved and close this issue accordingly.

**Comment by adrienthebo** [comment]
Since there have not been further updates on this issue I'm going to assume that this issue was resolved in r10k 1.5.0.
